implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java

```diff
@@ -327,10 +332,32 @@ public boolean rawValueEquals(String name, String expected) {
         return Objects.equals(expected, getRawValue(name));
     }
 
-    @Experimental("Extension to the original ConfigSource to allow retrieval of additional metadata on config lookup")
     public ConfigValue getConfigValue(String name) {
-        final ConfigValue configValue = configSources.getInterceptorChain().proceed(name);
-        return configValue != null ? configValue : ConfigValue.builder().withName(name).build();
+        return getConfigValue(name, true);
+    }
+
+    /**
+     * Return the {@link org.eclipse.microprofile.config.ConfigValue} for the specified property name from the underlying
+     * {@linkplain ConfigSource
+     * configuration source}. The lookup of the configuration is performed immediately, meaning that calls to
+     * {@link org.eclipse.microprofile.config.ConfigValue} will always yield the same results.
+     *
+     * @param name The configuration property name
+     * @param noSuchElementExceptionForbidden Indicates whether {@code NoSuchElementException} can be thrown or not.
+     * @return the resolved property value as a {@link org.eclipse.microprofile.config.ConfigValue}
+     * @throws NoSuchElementException in case the property value could not be expanded and
+     *         {@code noSuchElementExceptionForbidden} has been set to {@code false}.
+     */
+    public ConfigValue getConfigValue(String name, boolean noSuchElementExceptionForbidden) {
+        try {
+            final ConfigValue configValue = configSources.getInterceptorChain().proceed(name);
```

Comment:
I'm maybe thinking that we only need to thrown the `NoSuchElementException` when we cannot map to an actual type. In that case, it probably makes sense for the chain to not throw the Exception in the interceptor and throw it later when doing the conversion. I think it may be worth investigating. It should probably simplify the code.

Reply:
Yes it would be definitively more simpler but if we do so, we won't be able to throw a `NoSuchElementException` with a specific message indicating that the expression could not be expanded, is that ok with you?

Reply:
We need to check what the spec needs. We can still log an error, but we may loose some of the context to throw the proper exception later, unless we record it somehow in `ConfigValue`.